### PR TITLE
[IMP] mail: create messaging env outside of MessagingService

### PR DESCRIPTION
**Pull request by seb-odoo**
There should be no patch in tests, only a different env. This commit is a step
in that direction.

It also allows to use messaging env without MessagingService, which will be
needed on the public part.

Enterprise: https://github.com/odoo-dev/enterprise/pull/31

### CODE REVIEW

File: addons/mail/static/src/messaging/messaging_env.js [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+             * "Promise time".
+             */
+            await new Promise(resolve => setTimeout(resolve));
+            _addMessagingEntities(env);
+        });
+    }
+    Object.assign(env, { messagingCreatedPromise });
+
+    return {
+        removeMessaging: () => _removeMessagingEntities(env),
```
Comment (seb-odoo) [inline addons/mail/static/src/messaging/messaging_env.js]: can be removed?

File: addons/mail/static/src/messaging/test_utils.js [hunk 2]
```diff
@@ -34,6 +36,7 @@ const MockMailService = Class.extend({
         return BusService.extend({
             _beep() {}, // Do nothing
             _poll() {}, // Do nothing
+            _registerWindowUnload: function () {}, // Do nothing
```
Comment (seb-odoo) [inline addons/mail/static/src/messaging/test_utils.js]: waiting runbot before pushing but I will make this short notation

File: addons/mail/static/src/messaging/messaging_env.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    generateEntities,
+} = require('mail.messaging.entity.core');
+
+const { Store } = owl;
+const { EventBus } = owl.core;
+
+/**
+ * Adds messaging features to the given env.
+ * Note: some features will become available asynchronously, which will be
+ * notified with the resolution of `env.messagingCreatedPromise.
```
Comment (seb-odoo) [inline addons/mail/static/src/messaging/messaging_env.js]:
```suggestion
 * notified with the resolution of `env.messagingCreatedPromise`.
```

File: addons/mail/static/src/messaging/test_utils.js [hunk 4]
```diff
@@ -601,40 +634,21 @@ async function start(param0 = {}) {
...
-        'window.innerHeight': windowInnerHeight,
-        'window.innerWidth': windowInnerWidth,
-        'window.Notification': windowNotification,
+        debug = false,
+        env,
     } = param0;
+    const {
+        services = getMailServices({ env, hasChatWindow, debug }),
+    } = param0;
+    const session = services.messaging.env;
```
Comment (seb-odoo) [inline addons/mail/static/src/messaging/test_utils.js]: I think this should be it but somehow it doesn't break anything
```suggestion
    const { session } = services.messaging.env;
```

Comment (alexkuhn) [reply addons/mail/static/src/messaging/test_utils.js]: Yes, I think nothing breaks because code relies on session in env, not how session is passed to mocked environment. That means `web.session` is mis-configured during test, but not `this.env.session`.

File: addons/mail/static/tests/discuss_typing_notification_tests.js [hunk 5]
```diff
@@ -1,8 +1,8 @@
 odoo.define('mail.discuss_typing_notification_test', function (require) {
 "use strict";
 
-const { patchMessagingService } = require('mail.messaging.testUtils');
 var Timer = require('mail.model.Timer');
+const { getMailServices } = require('mail.messaging.testUtils');
```
Comment (alexkuhn) [inline addons/mail/static/tests/discuss_typing_notification_tests.js:4]: on top of `mail.model`

File: addons/mail/static/tests/systray/systray_messaging_menu_tests.js [hunk 6]
```diff
@@ -1072,11 +1068,10 @@ QUnit.test('messaging menu widget: click twice preview on slow message_fetch sho
             if (args.method === 'channel_minimize') {
                 // called to detach thread in chat window
                 // simulate longpolling response with new chat window state
-                const channelInfo = {
-                    ...self.data['mail.channel'].records[0],
+                const channelInfo = Object.assign({}, self.data['mail.channel'].records[0],{
                     is_minimized: true,
                     state: 'open',
-                };
+                });
```
Comment (alexkuhn) [inline addons/mail/static/tests/systray/systray_messaging_menu_tests.js:1074]: wouldn't make this change though. Chromium edge is being automatically deployed since almost 2 months now, and this syntax is nicer. Also this is code in test, so I would keep using spread operator on object.

File: addons/mail_bot/static/src/messaging/component/messaging_menu/messaging_menu_tests.js [hunk 7]
```diff
@@ -41,8 +41,12 @@ QUnit.test('rendering with OdooBot has a request (default)', async function (ass
             }
             return this._super(...arguments);
         },
-        'window.Notification': {
-            permission: 'default',
+        env: {
```
Comment (alexkuhn) [inline addons/mail_bot/static/src/messaging/component/messaging_menu/messaging_menu_tests.js:38]: should be defined above `mockRPC` (alphabetical order of obj keys)

File: addons/mail/static/src/messaging/test_utils.js [hunk 8]
```diff
@@ -213,12 +252,14 @@ function _useMessagingMenu(callbacks) {
...
  * @param {Object} [param0={}]
+ * @param {boolean} [env]
  * @param {boolean} [hasChatWindow]
+ * @param {boolean} [hasLegacyMail]
  * @param {boolean} [isDebug]
  * @returns {Object}
  */
-function getServices({ hasChatWindow, isDebug } = {}) {
-    return new MockMailService().getServices({ hasChatWindow, isDebug });
+function getMailServices({ env, hasChatWindow, hasLegacyMail, isDebug } = {}) {
```
Comment (alexkuhn) [inline addons/mail/static/src/messaging/test_utils.js:261]: any particular reason to rename? Don't think it's worth

File: addons/mail/static/src/messaging/test_utils.js [hunk 9]
```diff
@@ -213,12 +252,14 @@ function _useMessagingMenu(callbacks) {
 
 /**
  * @param {Object} [param0={}]
+ * @param {boolean} [env]
  * @param {boolean} [hasChatWindow]
+ * @param {boolean} [hasLegacyMail]
  * @param {boolean} [isDebug]
```
Comment (alexkuhn) [inline addons/mail/static/src/messaging/test_utils.js]: missing prefix `param0.`

File: addons/mail/static/src/messaging/entity/messaging/messaging_tests.js [hunk 10]
```diff
@@ -37,14 +37,8 @@ QUnit.module('Messaging', {
...
 
-    await this.start({
-        session: {
-            name: "Admin",
-            partner_id: 3,
-            partner_display_name: "Your Company, Admin",
-            uid: 2,
-        },
-    });
+    await this.start();
```
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/messaging/messaging_tests.js:40]: This test now depends on default session data of mocked messaging service.
Would be nicer to provide session data with `env.session`, so that data are explicit in this test.

File: addons/mail/static/src/messaging/component/discuss/discuss_tests.js [hunk 11]
```diff
@@ -5298,11 +5265,7 @@ QUnit.test('moderation: sent message accepted in moderated channel', async funct
         res_id: 20,
     }];
 
-    await this.start({
-        session: {
-            partner_id: 13,
-        },
-    });
+    await this.start();
```
Comment (alexkuhn) [inline addons/mail/static/src/messaging/component/discuss/discuss_tests.js:5268]: Again, would provide session data.
I think logic doesn't care of it, but this test simulates a message of itself being accepted, so message author partner id should match current partner id.

File: addons/mail/static/src/messaging/service/messaging_service/messaging_service.js [hunk 12]
```diff
@@ -1,50 +1,42 @@
...
-        }, this.messagingEnvExtension);
+        });
+
+        /**
+         * Messaging initialization.
+         */
+        const messagingInitializedPromise = this.env.messagingCreatedPromise.then(async () => {
+            // TODO FIXME The method uses service specific env keys so it can
+            // only be called after a service has properly set up those keys.
+            await this.env.messaging.start();
```
Comment (alexkuhn) [inline addons/mail/static/src/messaging/service/messaging_service/messaging_service.js:35]: So if I get it right, you need this because of uses of `trigger_up`/`call`/`rpc`/`do_action`/etc. a bit everywhere in code.

This is a bit hard to remove these legacy processes mainly because of mock server, and I assume because of this you may still need messaging service in front-end with livechat, as messaging entity is essential for the good working of messaging env.

I don't have a solution for this issue yet, and definitely none for the scope of this branch.
Still racking my brain on that from time to time, unfortunately not the highest priority for the moment as entity and concurrency issues are a bigger concern.

File: addons/mail/static/tests/systray/systray_messaging_menu_tests.js (hunk 6 above)
Comment (seb-odoo) [reply addons/mail/static/tests/systray/systray_messaging_menu_tests.js:1074]: Can't lint the rest of the file if there is a syntax error on it. Need to be consistent until we can actually change the rule everywhere.

File: addons/mail/static/src/messaging/test_utils.js (hunk 8 above)
Comment (seb-odoo) [reply addons/mail/static/src/messaging/test_utils.js:261]: One click "go to definition" instead of indirection nightmare.

File: addons/mail/static/src/messaging/component/discuss/discuss_tests.js (hunk 11 above)
Comment (seb-odoo) [reply addons/mail/static/src/messaging/component/discuss/discuss_tests.js:5268]: The test works even if it is not a message of the current user.

As long as it is a "pending message that is known" being accepted, the treatment should be the same.

It appears that in our business code, other users typically don't see pending messages, but there is no reason to test a particular case if the general case works as here.

Note that our `o-author` modifier name might be misleading, it actually means "non-moderator" (as per the condition to display it).

File: addons/mail/static/src/messaging/entity/messaging/messaging_tests.js (hunk 10 above)
Comment (seb-odoo) [reply addons/mail/static/src/messaging/entity/messaging/messaging_tests.js:40]: I think it's fine to test the default data in this case, this test has a very limited use (would be more useful as an integration test with data from a server than a unit test) and `currentPartner` is kind of magic relation itself too.

File: addons/mail/static/src/messaging/service/messaging_service/messaging_service.js (hunk 12 above)
Comment (seb-odoo) [reply addons/mail/static/src/messaging/service/messaging_service/messaging_service.js:35]: Well yes, this is the whole reason why we need messaging service as you explained to me.
We could get rid of the service it wasn't for those extra methods of legacy code.

We will need a public messaging service too for the same reason, which will be using the messaging_env the same way, except the default env will be the public one instead.

File: addons/mail/static/src/messaging/entity/messaging/messaging_tests.js (hunk 10 above)
Comment (seb-odoo) [reply addons/mail/static/src/messaging/entity/messaging/messaging_tests.js:40]: I made the origin of data very obvious by directly comparing to the corresponding session keys. The actual values are irrelevant, only the match is important.

File: addons/mail/static/tests/discuss_typing_notification_tests.js (hunk 5 above)
Comment (seb-odoo) [reply addons/mail/static/tests/discuss_typing_notification_tests.js:4]: done

File: addons/mail_bot/static/src/messaging/component/messaging_menu/messaging_menu_tests.js (hunk 7 above)
Comment (seb-odoo) [reply addons/mail_bot/static/src/messaging/component/messaging_menu/messaging_menu_tests.js:38]: done

File: addons/mail/static/src/messaging/test_utils.js (hunk 9 above)
Comment (seb-odoo) [reply addons/mail/static/src/messaging/test_utils.js]: done

File: addons/mail/static/src/messaging/test_utils.js (hunk 4 above)
Comment (seb-odoo) [reply addons/mail/static/src/messaging/test_utils.js]: I tried to pass the value correctly but it made bugs. Since it isn't required and it works fine without it I just removed it.

**Comment by seb-odoo** [comment]
Rebased. All suggestions done and/or commented. Check can be done again, and if ok just waiting on runbot to merge.

File: addons/mail/static/src/messaging/component/chat_window_manager/chat_window_manager_tests.js [hunk 13]
```diff
@@ -19,10 +19,11 @@ QUnit.module('ChatWindowManager', {
     beforeEach() {
         utilsBeforeEach(this);
         this.start = async params => {
-            let { env, widget } = await utilsStart(Object.assign({}, params, {
-                data: this.data,
+            let { env, widget } = await utilsStart(Object.assign({
                 hasChatWindow: true,
                 hasMessagingMenu: true,
+            }, params, {
+                data: this.data,
```
Comment (alexkuhn) [inline addons/mail/static/src/messaging/component/chat_window_manager/chat_window_manager_tests.js:26]: what's the motivation for this change?

File: addons/mail/static/src/messaging/test_utils.js [hunk 14]
```diff
@@ -59,27 +62,63 @@ const MockMailService = Class.extend({
...
+            autofetchPartnerImStatus: false,
+            disableAnimation: true,
+        });
+        return MessagingService.extend({
+            env: testEnv,
+        });
     },
     notification() {
-        return NotificationService;
+        return NotificationService.extend({});
```
Comment (alexkuhn) [inline addons/mail/static/src/messaging/test_utils.js]: no need empty obj, just `.extend()` should be enough

**Review by alexkuhn — APPROVED** [review]
looks good otherwise

File: addons/mail/static/src/messaging/component/chat_window_manager/chat_window_manager_tests.js (hunk 13 above)
Comment (seb-odoo) [reply addons/mail/static/src/messaging/component/chat_window_manager/chat_window_manager_tests.js:26]: To make sure `hasChatWindow` and `hasMessagingMenu` can be overriden by param on a test by test casis if necessary. Just making it more generic.
